lib/rules/no-unused-services.js

```diff
@@ -0,0 +1,203 @@
+}
+
+class Stack {
+  constructor() {
+    this.stack = new Array();
+  }
+  pop() {
+    return this.stack.pop();
+  }
+  push(item) {
+    this.stack.push(item);
+  }
+  peek() {
+    return this.stack.length > 0 ? this.stack[this.stack.length - 1] : undefined;
+  }
+}
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'disallow unused service injections',
+      category: 'Services',
+      recommended: false,
+      url:
+        'https://github.com/ember-cli/eslint-plugin-ember/tree/master/docs/rules/no-unused-services.md',
+    },
+    fixable: null,
+    schema: [],
+  },
+
+  getErrorMessage,
+
+  create(context) {
+    const classStack = new Stack();
+    const sourceCode = context.getSourceCode();
+
+    /**
+     * Gets the trailing comma token of the given node.
+     * If the trailing comma does not exist, this returns undefined.
+     * @param {ASTNode} node The given node
+     * @returns {Token|undefined} The trailing comma token or undefined
+     */
+    function getTrailingToken(node) {
+      const nextToken = sourceCode.getTokenAfter(node);
+      return types.isCommaToken(nextToken) ? nextToken : undefined;
+    }
+
+    /**
+     * Go through the current class and report any unused services
+     * @returns {void}
+     */
+    function reportInstances() {
+      const currentClass = classStack.pop();
+      const { services, uses } = currentClass;
+      if (!services || !uses) {
+        return;
+      }
+
+      for (const name of Object.keys(services)) {
+        if (!uses.has(name)) {
+          const node = services[name];
+          context.report({
+            node,
+            message: getErrorMessage(name),
+            suggest: [
+              {
+                desc: 'Remove the service injection if not used within the corresponding HBS file.',
+                fix: (fixer) => {
+                  const fixers = [fixer.remove(node)];
+                  if (types.isProperty(node)) {
+                    const trailingTokenNode = getTrailingToken(node);
+                    if (trailingTokenNode) {
+                      fixers.push(fixer.remove(trailingTokenNode));
+                    }
+                  }
+                  return fixers;
+                },
+              },
+            ],
+          });
+        }
+      }
+    }
+
+    return {
+      // Native JS class
+      ClassDeclaration(node) {
+        classStack.push({ node, services: {}, uses: new Set() });
+      },
+      CallExpression(node) {
+        if (emberUtils.isAnyEmberCoreModule(context, node)) {
+          // Classic class
+          classStack.push({ node, services: {}, uses: new Set() });
+        } else {
+          const currentClass = classStack.peek();
+          if (!currentClass) {
+            return;
+          }
+
+          if (types.isIdentifier(node.callee) && types.isThisExpression(node.arguments[0])) {
+            if (node.callee.name === 'get') {
```

Comment:
We can't actually assume these imported functions are named `get` or `getProperties`. We need to check the name they are actually imported under.

We can use this code from the `no-get` rule to store the names that these are imported under.

```js
ImportDeclaration(node) {
    if (node.source.value === '@ember/object') {
        importedGetName = importedGetName || getImportIdentifier(node, '@ember/object', 'get');
        importedGetPropertiesName =
            importedGetPropertiesName ||
            getImportIdentifier(node, '@ember/object', 'getProperties');
    }
},
```

That way, we could handle things like:

```js
import { get as customGetName } from '@ember/object';
```

Reply:
Ah yes, totally missed this! :) Thanks for pointing me to this - fixed.